### PR TITLE
Install to Module

### PR DESCRIPTION
**Pull request by nupplaphil**
Moved `mod/install.php` to `Module\Install`

- Using Status Code `204` instead of body `OK` for rewrite-testing (optional, I thought it's smoother with HTTP-Codes)
- Fixing issue #6038 
- Fixing issue that `local.ini.php` won't appear in finished text in case of not writable `config/`
- Fixing issue that `checkHtAccess()` used `$basepath` instead of `$baseurl`
- Fixing issue that `phppath` wasn't properly set during `checkAll()`

### CODE REVIEW

**Comment by nupplaphil** [comment]
- Tested with succeeding installation
- Tested with wrong permissions to `config/`

File: src/Module/Install.php [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+		// route: install/testrwrite
+		// $baseurl/install/testrwrite to test if rewrite in .htaccess is working
+		if ($a->getArgumentValue(1, '') == 'testrewrite') {
+			// Status Code 204 means that it worked without content
+			Core\System::httpExit(204);
+		}
+
+		// We overwrite current theme css, because during install we clould not have a working mod_rewrite
+		// so we could not have a css at all. Here we set a static css file for the install procedure pages
+		$a->setConfigValue('system', 'value', '../install');
```
Comment (MrPetovan) [inline src/Module/Install.php]: `value`?

File: src/Module/Install.php [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+		$a = self::getApp();
+
+		// route: install/testrwrite
+		// $baseurl/install/testrwrite to test if rewrite in .htaccess is working
+		if ($a->getArgumentValue(1, '') == 'testrewrite') {
+			// Status Code 204 means that it worked without content
+			Core\System::httpExit(204);
+		}
+
+		// We overwrite current theme css, because during install we clould not have a working mod_rewrite
```
Comment (MrPetovan) [inline src/Module/Install.php]: `clould`=> `may`

File: src/Module/Install.php [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+		// so we could not have a css at all. Here we set a static css file for the install procedure pages
```
Comment (MrPetovan) [inline src/Module/Install.php]: `could` => `may`

File: src/Module/Install.php [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+	/**
+	 * Creates the text for manual config
+	 *
+	 * @param App $a The global App
+	 *
+	 * @return string The manual config text
+	 */
+	private static function manualConfig($a) {
+		$data = htmlentities($a->data['txt'],ENT_COMPAT, 'UTF-8');
```
Comment (MrPetovan) [inline src/Module/Install.php]: Code standards: Please add a space after commas.

File: src/Module/Install.php [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+	}
+
+	/**
+	 * Creates the text for manual config
+	 *
+	 * @param App $a The global App
+	 *
+	 * @return string The manual config text
+	 */
+	private static function manualConfig($a) {
```
Comment (MrPetovan) [inline src/Module/Install.php]: Code standards: Please add a new line before class method opening brace.

File: src/Core/Install.php [hunk 6]
```diff
@@ -444,24 +443,23 @@ public function checkSmarty3()
...
 			$url = normalise_link($baseurl . "/install/testrewrite");
-			if ($fetchResult->getBody() != "ok") {
+			if ($fetchResult->getReturnCode() != 204) {
 				$fetchResult = Network::fetchUrlFull($url);
 			}
 
-			if ($fetchResult->getBody() != "ok") {
+			if ($fetchResult->getReturnCode() != 204) {
 				$status = false;
 				$help = L10n::t('Url rewrite in .htaccess is not working. Check your server configuration.');
```
Comment (MrPetovan) [inline src/Core/Install.php]: Shouldn't this be `Make sure you copied .htaccess-dist to .htaccess`?

File: src/Module/Install.php [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+	}
+
+	/**
+	 * Creates the text for the next steps
+	 *
+	 * @param App $a The global App
+	 *
+	 * @return string The text for the next steps
+	 */
+	private static function whatNext($a) {
```
Comment (MrPetovan) [inline src/Module/Install.php]: Code standards: Please add a new line before class method opening brace.

File: src/Module/Install.php [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+		$baseurl = $a->getBaseUrl();
+		return
+			L10n::t('<h1>What next</h1>')
+			."<p>".L10n::t('IMPORTANT: You will need to [manually] setup a scheduled task for the worker.')
```
Comment (MrPetovan) [inline src/Module/Install.php]: Code standards: Please wrap operators with spaces.

File: mod/install.php [hunk 9]
```diff
@@ -1,271 +0,0 @@
...
-	// $baseurl/install/testrwrite to test if rewite in .htaccess is working
-	if ($a->argc == 2 && $a->argv[1] == "testrewrite") {
-		echo "ok";
-		killme();
-	}
-
-	// We overwrite current theme css, because during install we could not have a working mod_rewrite
-	// so we could not have a css at all. Here we set a static css file for the install procedure pages
-
-	$a->setConfigValue('system', 'value', '../install');
```
Comment (nupplaphil) [inline mod/install.php]: @MrPetovan responding to https://github.com/friendica/friendica/pull/6041/files#r228921398

I copied it from here

File: src/Core/Install.php (hunk 6 above)
Comment (nupplaphil) [reply src/Core/Install.php]: We can add it too :-)

File: mod/install.php (hunk 9 above)
Comment (MrPetovan) [reply mod/install.php]: It doesn't look like it's used anywhere in the install, I didn't check the rest of the code but I would assume it's going to be the same.

Comment (nupplaphil) [reply mod/install.php]: I'll check it

**Comment by nupplaphil** [comment]
I refactored for using the `Core\Install` (now `Core\Installation`) to hold the state of the installation.

Both, the automatic installation and the web installation are now using the state of the installer for their decisions.

I tried to avoid using `$a->data[]` because as far as I can see, we use it as a container for every kind/type of stateful variable.

I'm not fully finished with testing, so plz don't merge it yet

File: src/Core/Console/AutomaticInstallation.php [hunk 10]
```diff
@@ -182,16 +178,30 @@ protected function doExecute()
 	}
 
 	/**
-	 * @param Install $install the Installer instance
+	 * @param Installer $install the Installer instance
+	 *
+	 * @return bool true if checks were successfully, otherwise false
 	 */
-	private function runBasicChecks(Install $install)
+	private function runBasicChecks(Installer $install)
```
Comment (MrPetovan) [inline src/Core/Console/AutomaticInstallation.php]: Please rename parameter to `$installer`.

**Comment by nupplaphil** [comment]
tested
- web installer with success
- web installer with not writable config
- web installer with wrong db credentials
- web installer with existing db
- automatic installer with the automated tests & one manual installation

**Comment by tobiasd** [comment]
Thanks for looking into it!

> - web installer with success

Have you tested the installation into a sub-directory as well? IIRC there were problems reported with such installations.

**Comment by nupplaphil** [comment]
Thx @tobiasd 

I had to add `$_SERVER['REQUEST_URI']` to make it work. But now I could install it with relative path. Registration & login worked too

**Comment by MrPetovan** [comment]
Part of #4090

**Comment by nupplaphil** [comment]
and it's a bugfix for #6038

**Comment by MrPetovan** [comment]
You already wrote that in the PR body!

**Comment by nupplaphil** [comment]
@MrPetovan Sorry, I meant that this isn't just an enhancement. It's a bugfix too :-) but never mind => deleted the old post. You can delete this one too if you like ;-)

**Pull request merged**